Composer/packages/form-dialogs/src/atoms/utils.ts

```diff
@@ -0,0 +1,289 @@
+  FormDialogPropertyPayload,
+  IntegerPropertyPayload,
+  NumberPropertyPayload,
+  RefPropertyPayload,
+  FormDialogPropertyKind,
+  StringPropertyPayload,
+  TypedPropertyPayload,
+} from 'src/atoms/types';
+import { generateId } from 'src/utils/base';
+
+export const getDefaultPayload = (kind: FormDialogPropertyKind) => {
+  switch (kind) {
+    case 'ref':
+      return <RefPropertyPayload>{ kind: 'ref' };
+    case 'string':
+      return <StringPropertyPayload>{ kind: 'string', entities: [] };
+    case 'number':
+      return <NumberPropertyPayload>{ kind: 'number', entities: [] };
+    case 'integer':
+      return <IntegerPropertyPayload>{ kind: 'integer', entities: [] };
+    default:
+      throw new Error(`Property type: "${kind}" is not supported!`);
+  }
+};
+
+const $refToRef = ($ref: string) => {
+  const [, ref] = $ref.match('template:(.*)');
+  return ref;
+};
+
+// eslint-disable-next-line @typescript-eslint/no-explicit-any
+const retrievePayload = (kind: FormDialogPropertyKind, payloadData: any, array = false): FormDialogPropertyPayload => {
+  if (array) {
+    return retrievePayload(payloadData.items.type || 'ref', payloadData.items);
+  }
+  switch (kind) {
+    case 'ref':
+      return <RefPropertyPayload>{ ref: $refToRef(payloadData.$ref) };
+    case 'string':
+      return <StringPropertyPayload>{ kind: 'string', entities: payloadData.$entities, enums: payloadData.enum };
+    case 'number':
+      return <NumberPropertyPayload>{
+        kind: 'number',
+        minimum: payloadData.minimum,
+        maximum: payloadData.maximum,
+      };
+    case 'integer':
+      return <IntegerPropertyPayload>{
+        kind: 'integer',
+        minimum: payloadData.minimum,
+        maximum: payloadData.maximum,
+      };
+    default:
+      throw new Error(`Property of type: ${kind} is not currently supported!`);
+  }
+};
+
+export const createSchemaStoreFromJson = (schemaName: string, jsonString: string) => {
+  const json = JSON.parse(jsonString);
+
+  const properties = json.properties || [];
+  const requiredArray = <string[]>(json.required || []);
+  const examplesRecord = <Record<string, string[]>>(json.$examples || {});
+
+  const propertyStores = Object.keys(properties).map((name) => {
+    const propertyData = properties[name];
+
+    const propertyType = propertyData?.type || 'ref';
+    const array = propertyType === 'array';
+    const payload = retrievePayload(propertyType, propertyData, array);
+    const kind = <FormDialogPropertyKind>(propertyType === 'array' ? payload.kind : propertyType);
+    const required = requiredArray.indexOf(name) !== -1;
+    const examples = examplesRecord[name] || [];
+
+    return {
+      id: generateId(),
+      name,
+      kind,
+      required,
+      examples,
+      array,
+      payload,
+    };
+  });
+
+  return { name: schemaName, properties: propertyStores };
+};
+
+const findFirstMissingIndex = (arr: number[], start: number, end: number): number => {
```

Comment:
i may have to pull you into interviews if i continue to see these clean common algorithms :)

Reply:
sounds good, Geoff knows ;)